### PR TITLE
feat(document-loader-mcp-server): add Document Loader MCP Server

### LINKED ISSUE #1157: RFC: Add new Document Loading MCP Server
### Is this related to an existing feature request or issue?

No

### Summary

We propose contributing the Document Loader MCP Server to AWS Labs. This server enables AI applications and tools to interact with various document formats using natural language for content extraction and analysis. The server provides a unified interface for reading PDFs, Word documents, Excel spreadsheets, PowerPoint presentations, and images, converting them into LLM formats while maintaining content structure and accuracy.

The MCP server implements 5 core tools covering document processing from basic text extraction to advanced multi-format support, with robust error handling and comprehensive testing infrastructure.

### Document Analysis and Content Extraction
- Extract and analyze content from business documents, reports, and presentations
- Enable AI tools to process mixed-format document collections
- Support automated document review and summarization workflows
- Convert legacy documents into searchable, LLM-processable formats

### Knowledge Management and RAG Applications
- Prepare documents for vector embedding and semantic search
- Extract structured content for knowledge base construction
- Support multi-format document ingestion pipelines
- Enable cross-format content correlation and analysis

### Business Process Automation
- Automate invoice, contract, and report processing
- Extract data from spreadsheets for analysis and reporting
- Process presentation content for meeting summaries
- Support compliance and audit document review

### Multi-modal AI Workflows
- Combine text extraction with image analysis capabilities
- Support documents containing both text and visual elements
- Enable comprehensive document understanding across formats
- Facilitate cross-format content comparison and validation

With the Document Loader MCP Server, users can have conversations like:

**User**: "Extract the text content from this quarterly report PDF and summarize the key findings"
**AI Assistant**: "I've extracted the content from the PDF. The quarterly report shows revenue growth of 15% year-over-year, with key highlights including..."

**User**: "Read this Excel spreadsheet and tell me the total sales by region"
**AI Assistant**: "Based on the spreadsheet data, here are the total sales by region: North America: $2.3M, Europe: $1.8M, Asia-Pacific: $1.2M..."

**User**: "What are the main topics covered in this PowerPoint presentation?"
**AI Assistant**: "The presentation covers 4 main topics: Market Analysis, Product Strategy, Financial Projections, and Implementation Timeline..."

### What We're Building

A Python-based MCP server that exposes 5 comprehensive tools for AI assistants:

**Document Reading Tools**
- **read_pdf** - Extract text content from PDF files using pdfplumber with page-by-page processing
- **read_docx** - Convert Word documents to markdown using markitdown for structured output
- **read_xlsx** - Extract and convert Excel spreadsheet data to markdown format
- **read_pptx** - Process PowerPoint presentations and convert slide content to markdown

**Visual Content Tools**
- **read_image** - Load and display images for AI analysis with support for multiple formats

### Technical Approach

- **Core Framework**: Built on FastMCP for robust MCP protocol implementation
- **PDF Processing**: Uses pdfplumber for accurate text extraction with page-level granularity
- **Office Document Processing**: Leverages markitdown library for consistent markdown conversion
- **Image Handling**: Supports PNG, JPG, JPEG, GIF, BMP, TIFF, WEBP formats via FastMCP Image utilities
- **Error Handling**: Comprehensive error handling with descriptive error messages
- **Testing Infrastructure**: Automated test generation using reportlab, python-docx, openpyxl, and python-pptx
- **Packaging**: Distributed as Python wheel for easy installation and deployment

### Integration

Works with any MCP-compatible AI assistant:
- **Kiro** - Primary development and testing environment
- **Claude Desktop** - Full MCP protocol support
- **Amazon Q Developer** - Enterprise AI assistant integration
- **GitHub Copilot** - Development workflow integration
- **Custom AI Applications** - Any MCP-compatible system

### Document Modification and Creation
- Document editing or modification capabilities
- Document generation or template creation
- Watermarking or digital signature operations
- **Rationale**: Focus on read-only operations for security and simplicity

### Advanced Document Processing
- OCR (Optical Character Recognition) for scanned documents
- Complex table extraction and formatting preservation
- Document comparison and diff generation
- **Rationale**: These features require specialized libraries and significantly increase complexity

### Cloud Storage Integration
- Direct integration with cloud storage services
- Remote document fetching and caching
- Authentication for cloud services
- **Rationale**: Maintains focus on local file processing and avoids authentication complexity

### Real-time Document Processing
- Document change monitoring and real-time updates
- Streaming document processing for large files
- Concurrent multi-document processing
- **Rationale**: Batch processing model is simpler and more reliable for MCP use cases

### Large Document Processing
**Challenge**: Memory usage and processing time for large documents (>100MB PDFs, complex spreadsheets)
**Mitigation**: 
- Implemented page-by-page PDF processing to manage memory usage
- Added comprehensive error handling for timeout and memory issues
- Clear documentation of recommended file size limits
- Streaming processing where possible

### Format Compatibility and Edge Cases
**Challenge**: Handling corrupted files, password-protected documents, and unusual formatting
**Mitigation**:
- Robust error handling with descriptive error messages
- Graceful degradation for partially readable documents
- Clear documentation of supported formats and limitations
- Comprehensive test suite covering edge cases

### Dependency Management and Compatibility
**Challenge**: Managing complex dependencies (pdfplumber, markitdown, office libraries) across different platforms
**Mitigation**:
- Pinned dependency versions in pyproject.toml for reproducible builds
- Comprehensive testing across document types
- Wheel-based distribution for consistent deployment
- Clear documentation of system requirements

### Performance and Scalability
**Challenge**: Processing multiple large documents efficiently without blocking the MCP server
**Mitigation**:
- Efficient libraries chosen for each format (pdfplumber, markitdown)
- Memory-conscious processing approaches
- Clear performance expectations in documentation
- Timeout handling for long-running operations

### Core Dependencies
- **fastmcp>=2.11.1** - MCP server framework and protocol implementation
- **pdfplumber>=0.10.0** - Robust PDF text extraction with layout awareness
- **markitdown[docx,xlsx,pptx]>=0.1.2** - Multi-format document conversion to markdown

### Development and Testing Dependencies
- **pytest>=7.0.0** - Testing framework with async support
- **reportlab>=4.0.0** - PDF generation for testing
- **python-docx>=1.1.0** - DOCX generation for testing
- **openpyxl** - Excel file generation and testing
- **python-pptx** - PowerPoint generation for testing

### Alternative solutions

```markdown
### Existing Document Processing Libraries
**Alternative**: Direct use of pdfplumber, python-docx, openpyxl individually
**Limitations**: 
- Requires separate integration for each format
- No unified interface for AI assistants
- Manual error handling and format detection
- No MCP protocol support
```

### PR DESCRIPTION
**Pull request by hvital**
# Add Document Loader MCP Server

Add MCP server for document parsing and content extraction supporting:
- PDF text extraction via pdfplumber
- Word document processing (DOCX/DOC) via markitdown
- Excel spreadsheet parsing (XLSX/XLS) via markitdown
- PowerPoint presentation processing (PPTX/PPT) via markitdown
- Image loading for various formats (PNG, JPG, GIF, BMP, TIFF, WEBP)

Fixes #1157

### Changes

This PR introduces the **Document Loader MCP Server** (`awslabs.document-loader-mcp-server`), a new Model Context Protocol server that enables AI assistants to parse and extract content from various document formats.

**Key Components Added:**
- **Core Package**: MCP server implementation with FastMCP framework
- **Document Processing Tools**: 5 tools for different document types
- **Testing**: Test suite with automated sample document generation (6 tests)
- **Docker Support**: Container deployment with health checks
- **Documentation**: README with installation instructions and usage examples

**Files Added:**
- `src/document-loader-mcp-server/` - Package directory with 17 files
- Updated main `README.md` with Document Loader entry
- Updated `.github/CODEOWNERS` with maintainer assignment

### User Experience

**Before**: Users had no standardized way to extract content from documents (PDF, Office files, images) for AI assistant processing through MCP protocol.

**After**: Users can:
1. Install via `uvx awslabs.document-loader-mcp-server@latest`
2. Configure in their MCP client (Cursor, VS Code, Amazon Q Developer CLI)
3. Use 5 document processing tools:
   - `read_pdf` - Extract text from PDF files with page-by-page processing
   - `read_docx` - Convert Word documents to markdown
   - `read_xlsx` - Parse Excel spreadsheets to markdown
   - `read_pptx` - Extract PowerPoint content to markdown
   - `read_image` - Load images for LLM analysis

**Example Usage:**
```json
{
  "mcpServers": {
    "awslabs.document-loader-mcp-server": {
      "command": "uvx",
      "args": ["awslabs.document-loader-mcp-server@latest"],
      "env": {"FASTMCP_LOG_LEVEL": "ERROR"},
      "disabled": false,
      "autoApprove": []
    }
  }
}
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] I have reviewed the [contributing guidelines](https://github.com/awslabs/mcp/blob/main/CONTRIBUTING.md)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented

Is this a breaking change? (Y/N) **N**

**RFC issue number**: #1157

Checklist:
* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the [project license](https://github.com/awslabs/mcp/blob/main/LICENSE).

### CODE REVIEW

File: src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        return result.text_content
+
+    except FileNotFoundError:
+        return f'Error: Could not find {file_type} at {file_path}'
+    except Exception as e:
+        return f'Error reading {file_type} {file_path}: {str(e)}'
+
+
+@mcp.tool()
+def read_pdf(file_path: str) -> str:
```
Comment (krokoko) [inline src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py:53]: For tools here (in the whole server), could we use a Pydantic model for request/response validation and type safety ? 
For instance something like:

```
from pydantic import BaseModel, Field
from typing import Optional

class DocumentReadResponse(BaseModel):
    """Response from document reading operations."""
    status: str
    content: str
    file_path: str
    error_message: Optional[str] = None

@mcp.tool(name='read_pdf')
async def read_pdf(
    ctx: Context,
    file_path: str = Field(..., description="Path to the PDF file to read")
) -> DocumentReadResponse:
```

File: src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        return text_content.strip()
+
+    except FileNotFoundError:
+        return f'Error: Could not find PDF file at {file_path}'
+    except Exception as e:
+        return f'Error reading PDF file {file_path}: {str(e)}'
+
+
+@mcp.tool()
+def read_docx(file_path: str) -> str:
```
Comment (krokoko) [inline src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py:82]: For all tools: any reason to not follow the async programming guideline ? For instance like this https://github.com/awslabs/mcp/blob/main/src/aws-documentation-mcp-server/awslabs/aws_documentation_mcp_server/server_aws.py#L78
In the guidelines here: https://github.com/awslabs/mcp/blob/main/DESIGN_GUIDELINES.md#asynchronous-programming

File: src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+def read_pdf(file_path: str) -> str:
+    """Extract text content from a PDF file (*.pdf).
+
+    Args:
+        file_path (str): Path to the PDF file to read
+
+    Returns:
+        str: Extracted text content from the PDF or error message
+    """
+    try:
```
Comment (krokoko) [inline src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py:62]: for each tool: is there any additional validation from a security perspective we could do ? 
For instance, have a file size limit check, path traversal attacks,...?

File: src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py (hunk 1 above)
Comment (krokoko) [inline src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py:53]: For each tool: Include the Context parameter for error reporting

File: src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py (hunk 3 above)
Comment (krokoko) [reply src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py:62]: For instance, something like:

```
# Add to consts.py
MAX_FILE_SIZE = 50 * 1024 * 1024  # 50MB limit
ALLOWED_EXTENSIONS = {'.pdf', '.docx', '.doc', '.xlsx', '.xls', '.pptx', '.ppt', '.png', '.jpg', '.jpeg', '.gif', '.bmp', '.tiff', '.tif', '.webp'}
ALLOWED_PATHS = ['/tmp',]  # Restrict to safe directories, could be set through an env var maybe ? or use mcp roots but not sure how many clients support it, great use case for it though

# Add input validation
def validate_file_path(file_path: str) -> bool:
    """Validate file path for security."""
    path = Path(file_path).resolve()
    
    # Check file size
    if path.stat().st_size > MAX_FILE_SIZE:
        raise ValueError(f"File too large: {path.stat().st_size} bytes")
    
    # Check extension
    if path.suffix.lower() not in ALLOWED_EXTENSIONS:
        raise ValueError(f"Unsupported file type: {path.suffix}")
    
    # Check path traversal
    for allowed_path in ALLOWED_PATHS:
        if str(path).startswith(allowed_path):
            return True
    
    raise ValueError("File path not in allowed directories")
```

File: src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py (hunk 1 above)
Comment (krokoko) [reply src/document-loader-mcp-server/awslabs/document_loader_mcp_server/server.py:53]: More info: https://gofastmcp.com/servers/context

**Comment by krokoko** [comment]
Closing this one as https://github.com/awslabs/mcp/pull/1379 is the new PR for this server
